### DISCUSSION TITLE
A chance Microsoft give you to fully support the latest Windows API

### DISCUSSION THREAD

**Discussion by nsgnkhibdk2cls0f** [modula3/cm3#1142] [General]
Many other languages build their Windows API interfaces using this Microsoft tool:

https://github.com/microsoft/win32metadata

It's up you to decide to make Modula-3 up to speed with other languages or not.

**Comment by dragisha** [comment]
I, for one, do not care about Microsoft APIs for several reasons. One - I am *nix person, developing for Linux and using macOS as my workstation.

Also, I like portability. So when I am developing something with UI I will use Gtk so I can run this on Windows, Linux, macOS and get same results. 

For people happy to lock themselves into Windows Modula-3 is absolutely wrong choice. Use Visual something, get on with .NET something… Do not make your life hard with Modula-3!

**Reply by mikanystrom-intel** [reply]
I can only agree 100%.

Of course I don't object if people want to use Modula-3 on Windows, but if they do, I think it's up to them to figure out the bindings...

**Reply by nsgnkhibdk2cls0f** [reply]
@dragisha Gtk? Does CM3 has any Gtk bindings available? Btw, even with Gtk, the expected way to create Gtk bindings is GObject Introspection (generating bindings automatically via the GIR files):

https://gi.readthedocs.io/en/latest/

https://gitlab.gnome.org/GNOME/gobject-introspection

**Reply by ghost** [reply]
> For people happy to lock themselves into Windows Modula-3 is absolutely wrong choice.

ADW Software Modula-2 (full Windows API support): https://www.modula2.org/adwm2/

It's no reasons Modula-3 can't be good with Windows programming.

**Reply by ghost** [reply]
@nsgnkhibdk2cls0f Yep, Microsoft actually learned the idea from Gtk.

**Reply by dmuysers** [reply]
>> Also, I like portability. So when I am developing something with UI I will use Gtk so I can run this on Windows, Linux, macOS and get same results.
There is also IUP from Tecgraf
https://www.tecgraf.puc-rio.br/iup/

**Reply by nsgnkhibdk2cls0f** [reply]
I know about IUP. FreeBasic has full bindings for IUP and other projects developed by scuri like IM, CD. FreeBasic also has many text editors and IDEs based on IUP, some written in FreeBasic itself. IUP is fairly well known on the FreeBasic community.

The problem is, does IUP has any working binding for CM3? You all mentioned a library without showing if they have a working binding for CM3 or not. If they are not, they are pretty much useless for a CM3 user unless he could make a CM3 binding himself. Both Gtk and IUP are complex pieces of C code very difficult to create a CM3 binding for.

**Reply by nsgnkhibdk2cls0f** [reply]
It seems both Gtk and IUP have no working CM3 binding. Windows API is different. Even though the current interface is incomplete, it's already there. What I proposed on this thread is simply make use of Microsoft's tool to automatically generate CM3 interface for Windows API, thus updating your dated version to the latest version on par with Microsoft.

**Reply by nsgnkhibdk2cls0f** [reply]
https://github.com/modula3/cm3/issues/1124

**Reply by nsgnkhibdk2cls0f** [reply]
https://github.com/modula3/cm3/issues/1062

**Comment by VictorMiasnikov** [comment]
How about using m2tom3 converter?

( I wrote more details inside another discussion)

**Reply by 1997tn** [reply]
ADW Modula-2's WinAPI headers may have licensing problems. I don't know if it's legal to just convert their headers and use with CM3. I'm not sure about the licensing of it, if you are allowed to do such a thing or not, ADW Modula-2 only said it's freeware (not even open source).

**Reply by VictorMiasnikov** [reply]
} ADW Modula-2's WinAPI headers may have licensing problems.

 I.e. we can read this phrase as
 ADW Modula-2's WinAPI headers may does not have licensing problems.
 
 
 IMHO, the simplest way is directly ask owner of ADW Modula-2.

**Reply by 1997tn** [reply]
This is the address listed on env.chm part of ADW Modula-2 distro: fons.dw@pythagoras.net

modula2.org has a feedback form: https://www.modula2.org/feedback.php

Your message will be go to modula2.org's site owner first before to ADW Modula-2 if you use this feedback form.

**Reply by 1997tn** [reply]
@VictorMiasnikov Download this and unpack: https://github.com/1997tn/modula2/raw/main/ADW%20Software%20Modula-2.zip

You could try it yourself. I tried. It doesn't work.

**Comment by VictorMiasnikov** [comment]
> This is the address listed on env.chm part of ADW Modula-2 distro: (...)

In case when one distributive of Modula-2 has non-free license, then we can use source code of another distributive.

 We can try, for example, XDS Oberon-2/Modula-2

**Reply by 1997tn** [reply]
The problem is, does XDS Oberon-2/Modula-2 has a fairly complete WinAPI headers like ADW Modula-2?

**Comment by VictorMiasnikov** [comment]
> } Download this (...)

Thank You, I have 2 - 3 binary distributive of ADW Modula-2.

 (  And I have successfully compiled XDS Modula-2/Oberon-2 )

P.S.

 We even can try early ( with non-broken Cygwin target) GNU Modula-2

**Reply by 1997tn** [reply]
So you have ADW Modula-2 already. ADW Modula-2 installer can't be extracted to get headers like other installers. I have to install it and zip it in case someone wanted to get the headers they don't have to install ADW Modula-2 like me. It seems not needed anymore, I will delete this repository. Btw, are you sure XDS Modula-2/Oberon-2 has a fairly complete WinAPI headers like ADW Modula-2?

**Reply by 1997tn** [reply]
GNU Modula-2 is very bad at Windows support. It's not even work natively without Cygwin (currently it's also broken on Cygwin). Why do you expect it has WinAPI headers?

**Comment by VictorMiasnikov** [comment]
> } but only discussing about how to steal WinAPI headers from a Modula-2 distro?

 Personal for me, "WinAPI" is only example.

 Try use "SWIG for Modula-3" and ( or) m2tom3 ( Modula-2 to Modula-3 translator / converter) and public bad or good results.

( if You want, of course.

Thank! )
